src/librustc_typeck/check/mod.rs

```diff
@@ -4918,18 +4918,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             format!("{} type parameter{}", n, if n == 1 { "" } else { "s" })
         };
 
-        // Check provided type parameters.
-        let type_defs = segment.map_or(&[][..], |(_, generics)| {
-            if generics.parent.is_none() {
-                &generics.types[generics.has_self as usize..]
-            } else {
-                &generics.types
-            }
-        });
-        let required_len = type_defs.iter().take_while(|d| !d.has_default).count();
-        if types.len() > type_defs.len() {
-            let span = types[type_defs.len()].span;
-            let expected_text = count_type_params(type_defs.len());
+        // Check provided parameters.
+        let (ty_non_def_req_len, ty_req_len, lt_req_len) =
+            segment.map_or((0, 0, 0), |(_, generics)| {
+                let param_counts = generics.param_counts();
+
+                let type_params_offset
+                    = (generics.parent.is_none() && generics.has_self) as usize;
+                let type_params = param_counts.types - type_params_offset;
+                let type_params_without_defaults = {
+                    let mut count = 0;
+                    for param in generics.params.iter() {
+                        if let ty::GenericParamDefKind::Type(ty) = param.kind {
+                            if !ty.has_default {
+                                count += 1
+                            }
+                        }
+                    }
+                    count
+                };
+                let type_params_barring_defaults =
+                    type_params_without_defaults - type_params_offset;
+
+                (type_params_barring_defaults, type_params, param_counts.lifetimes)
+            });
+
+        if types.len() > ty_req_len {
```

Comment:
I'd call this `ty_max_len` and `ty_non_def_req_len` can be just `ty_req_len` ("required" as in you can't supply *less*).